pipeline_dp/dataframes.py

```diff
@@ -0,0 +1,189 @@
+# Copyright 2023 OpenMined.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Computing DP aggregations on (Pandas, Spark, Beam) Dataframes."""
+
+from dataclasses import dataclass
+from typing import Any, Dict, Iterable, List, Optional
+
+import pipeline_dp
+
+
+@dataclass
+class Columns:
+    privacy_key: str
+    partition_key: str
+    value: Optional[str]
+
+
+@dataclass
+class ContributionBounds:
+    max_partitions_contributed: Optional[int] = None
+    max_contributions_per_partition: Optional[int] = None
+    min_value: Optional[float] = None
+    max_value: Optional[float] = None
+
+
+@dataclass
+class Budget:
+    epsilon: float
+    delta: float = 0
+
+    def __post_init__(self):
+        if self.epsilon <= 0:
+            raise ValueError(f"epsilon must be positive, not {self.epsilon}.")
+        if self.delta < 0:
+            raise ValueError(f"delta must be non-negative, not {self.delta}.")
+
+
+class Query:
+    """DP Query on DataFrames.
+
+    Warning: Don't create it directly, use QueryBuilder().
+    """
+
+    def __init__(self, df, columns: Columns, metrics: Dict[pipeline_dp.Metric,
+                                                           List[str]],
+                 contribution_bounds: ContributionBounds, public_keys):
+        self._df = df
+        self._columns = columns
+        self._metrics = metrics
+        self._contribution_bounds = contribution_bounds
+        self._public_partitions = public_keys
+
+    def run_query(self,
+                  budget: Budget,
+                  noise_kind: Optional[pipeline_dp.NoiseKind] = None):
```

Comment:
should be Laplace by default? we can't run without noise.

Reply:
I though initially to make it automatic (if None is set). Let's go simple and set it to Laplace